Python/ast.c

```diff
@@ -326,12 +326,234 @@ validate_expr(expr_ty exp, expr_context_ty ctx)
+                    }
+                }
+                else {
+                    PyErr_SetString(
+                        PyExc_ValueError,
+                        "unexpected key type for a mapping pattern");
+                    return 0;
+                }
+            }
+            return 1;
+        case List_kind:
+        case Tuple_kind:
+            values = pattern->kind == List_kind ? pattern->v.List.elts
+                                                : pattern->v.Tuple.elts;
+            length = asdl_seq_LEN(values);
+
+            expr_ty element;
+            for (i = 0; i < length; i++) {
+                element = asdl_seq_GET(values, i);
+                if (element->kind == Starred_kind) {
+                    if (element->v.Starred.value->kind == Name_kind) {
+                        if (!validate_expr(element->v.Starred.value, Store)) {
+                            return 0;
+                        }
+                    }
+                    else {
+                        PyErr_SetString(PyExc_ValueError,
+                                        "the single star pattern's target can "
+                                        "only be Name");
+                        return 0;
+                    }
+                }
+                else if (!validate_pattern(element)) {
+                    return 0;
+                }
+            }
+            return 1;
+        case UnaryOp_kind:
+            if (pattern->v.UnaryOp.op != USub) {
+                PyErr_SetString(PyExc_ValueError,
+                                "only USub operator is allowed for UnaryOp "
+                                "nodes inside of a pattern");
+                return 0;
+            }
+            if (pattern->v.UnaryOp.operand->kind == Constant_kind) {
+                literal = pattern->v.UnaryOp.operand->v.Constant.value;
+                if (PyComplex_CheckExact(literal) ||
+                    PyFloat_CheckExact(literal) ||
+                    PyLong_CheckExact(literal)) {
+                    return 1;
+                }
+                else {
+                    PyErr_SetString(PyExc_ValueError,
+                                    "only numbers are allowed as an operand "
+                                    "to UnaryOp nodes inside of a pattern");
+                    return 0;
+                }
+            }
+            PyErr_SetString(PyExc_ValueError,
+                            "only Constant nodes are allowed as an operand to "
+                            "the UnaryOp node");
+            return 0;
+        case MatchAs_kind:
+            return validate_pattern(pattern->v.MatchAs.pattern) &&
+                   validate_name(pattern->v.MatchAs.name);
+        case Name_kind:
+            if (pattern->v.Name.ctx == Store) {
+                return 1;
+            }
+            PyErr_SetString(PyExc_ValueError,
+                            "Expecting store context for name pattern");
+            return 0;
```

Comment:
Need to validate the actual name as well:

```suggestion
            return validate_expr(pattern, Store));
```